java/fury-core/src/main/java/org/apache/fury/Fury.java

```diff
@@ -1220,6 +1240,47 @@ public Object deserializeJavaObjectAndClass(FuryReadableChannel channel) {
     return deserializeJavaObjectAndClass(buf);
   }
 
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> T copy(T obj) {
+    if (Objects.isNull(obj)) {
+      return null;
+    }
+    try {
+      if (obj instanceof FuryCopyable) {
+        return (T) ((FuryCopyable<?>) obj).copy(this);
+      } else {
+        return (T) classResolver.getOrUpdateClassInfo(obj.getClass()).getSerializer().copy(obj);
+      }
+    } catch (StackOverflowError e) {
+      throw processCopyStackOverflowError(e);
+    } finally {
+      if (copyRefTracking && copyDepth == 0) {
+        resetCopy();
+      }
+    }
+  }
+
+  /**
+   * Record the mapping between the object currently being copied and the newly generated object
+   * after copying.
+   *
+   * @param originObj the origin object instance
+   * @param newObj the new object instance
+   */
+  public void copyReference(Object originObj, Object newObj) {
+    if (Objects.nonNull(originObj) && Objects.nonNull(newObj)) {
```

Comment:
Why not just write `originObj != null && newObj != null`? And for ref tracking, I think checking originObj is enough